### PR TITLE
[myStrom] Initial Contribution

### PR DESCRIPTION
**Pull request by pail23**
This is an initial version of a binding for the myStrom smart plug.

### CODE REVIEW

**Review by Hilbrand — CHANGES_REQUESTED** [review]
Thanks for you contribution. The code looks great. I've added some minor improvements and one about caching data.

File: bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandler.java [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+                if (command instanceof OnOffType && CHANNEL_SWITCH.equals(channelUID.getId())) {
+                    sendHttpGet("relay?state=" + (command == OnOffType.ON ? "1" : "0"));
+                    scheduler.schedule(() -> {
+                        pollDevice();
+                    }, 500, TimeUnit.MILLISECONDS);
+                }
+
+            }
+        } catch (MyStromException e) {
+            logger.error(COMMUNICATION_ERROR, e);
```
[the remark this answers is not in the document]
Comment (pail23) [reply bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandler.java]: changed to warn

File: bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandler.java [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+    private void pollDevice() {
+        try {
+            String returnContent = sendHttpGet("report");
+            MyStromReport report = gson.fromJson(returnContent, MyStromReport.class);
+            updateState(CHANNEL_SWITCH, report.relay ? OnOffType.ON : OnOffType.OFF);
+            updateState(CHANNEL_POWER, QuantityType.valueOf(report.power, WATT));
+            updateState(CHANNEL_TEMPERATURE, QuantityType.valueOf(report.temperature, CELSIUS));
+            updateStatus(ThingStatus.ONLINE);
+        } catch (MyStromException e) {
+            logger.error(COMMUNICATION_ERROR, e);
```
[the remark this answers is not in the document]
Comment (pail23) [reply bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandler.java]: changed to debug

File: bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandler.java [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+        logger.debug("Start initializing!");
+        config = getConfigAs(MyStromConfiguration.class);
+        hostname = config.hostname;
+
+        pollingJob = scheduler.scheduleWithFixedDelay(this::pollDevice, DEFAULT_WAIT_BEFORE_INITIAL_REFRESH,
+                config.refresh, TimeUnit.SECONDS);
+
+        updateStatus(ThingStatus.UNKNOWN);
+
+        scheduler.execute(this::pollDevice);
```
[the remark this answers is not in the document]
Comment (pail23) [reply bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandler.java]: removed this line and change to 0 delay in the previous call

File: bundles/org.openhab.binding.mystrom/src/main/resources/ESH-INF/thing/thing-types.xml [hunk 4]
```diff
@@ -0,0 +1,53 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<thing:thing-descriptions bindingId="mystrom"
+	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xmlns:thing="https://openhab.org/schemas/thing-description/v1.0.0"
+	xsi:schemaLocation="https://openhab.org/schemas/thing-description/v1.0.0 https://openhab.org/schemas/thing-description-1.0.0.xsd">
+
+	<!-- Sample Thing Type -->
```
[the remark this answers is not in the document]
Comment (pail23) [reply bundles/org.openhab.binding.mystrom/src/main/resources/ESH-INF/thing/thing-types.xml]: done

File: bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandler.java [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+    private ScheduledFuture<?> pollingJob;
+
+    private final Gson gson = new Gson();
+
+    public MyStromHandler(Thing thing, HttpClient httpClient) {
+        super(thing);
+        this.httpClient = httpClient;
+    }
+
+    private static class MyStromReport {
```
[the remark this answers is not in the document]
Comment (pail23) [reply bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandler.java]: Moved the class to the top

File: bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandler.java [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+        public float power;
+        public boolean relay;
+        public float temperature;
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        try {
+            if (command instanceof RefreshType) {
+                pollDevice();
```
[the remark this answers is not in the document]
Comment (pail23) [reply bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandler.java:88]: Added the expiry cache

**Comment by pail23** [comment]
Many thanks @Hilbrand for the fast review! I have implemented your comment. Please let me know in case I can further improve the code.

**Review by Hilbrand — APPROVED** [review]
LGTM

File: bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandler.java [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+            } else {
+                if (command instanceof OnOffType && CHANNEL_SWITCH.equals(channelUID.getId())) {
+                    sendHttpGet("relay?state=" + (command == OnOffType.ON ? "1" : "0"));
+                    scheduler.schedule(this::pollDevice, 500, TimeUnit.MILLISECONDS);
+                }
+
+            }
+        } catch (MyStromException e) {
+            logger.warn(COMMUNICATION_ERROR, e);
```
Comment (fwolter) [inline bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandler.java:94]: It seems a bit unreasonable logging the stack trace when the network fails. You could simply log the exception's message.

File: bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandler.java [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+
+    private @Nullable MyStromReport getReport() {
+        try {
+            String returnContent = sendHttpGet("report");
+            MyStromReport report = gson.fromJson(returnContent, MyStromReport.class);
+            updateStatus(ThingStatus.ONLINE);
+            return report;
+        } catch (MyStromException e) {
+            logger.debug(COMMUNICATION_ERROR, e);
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, COMMUNICATION_ERROR);
```
Comment (fwolter) [inline bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandler.java:106]: You could use the exception's message to fill the last argument of `updateStatus()`. The log message would be obsolete, then, as the status update is already logged.

File: bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandler.java [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+            updateState(CHANNEL_TEMPERATURE, QuantityType.valueOf(report.temperature, CELSIUS));
+            updateStatus(ThingStatus.ONLINE);
+        } else {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, COMMUNICATION_ERROR);
+        }
+    }
+
+    @Override
+    public void initialize() {
+        logger.debug("Start initializing!");
```
Comment (fwolter) [inline bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandler.java:125]: Can this message be replaced by using the debugger or by increasing the framework's log level? See point 4 https://www.openhab.org/docs/developer/guidelines.html#f-logging

File: bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandler.java [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+        config = getConfigAs(MyStromConfiguration.class);
+        hostname = config.hostname;
```
Comment (fwolter) [inline bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandler.java:127]: If you make `config` a local variable, the compiler warning would be gone.

File: bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandler.java [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+
+        updateStatus(ThingStatus.UNKNOWN);
+        pollingJob = scheduler.scheduleWithFixedDelay(this::pollDevice, 0, config.refresh, TimeUnit.SECONDS);
+
+        logger.debug("Finished initializing!");
```
Comment (fwolter) [inline bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandler.java:132]: Can this message be replaced by using the debugger or by increasing the framework's log level?

File: bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandler.java [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+     * @param url The URL to send a GET request to.
+     * @return String contents of the response for the GET request.
+     * @throws Exception
+     */
+    public String sendHttpGet(String action) throws MyStromException {
+
+        String url = "http://" + hostname + "/" + action;
+        ContentResponse response = null;
+        try {
+            response = httpClient.newRequest(url).method(HttpMethod.GET).send();
```
Comment (fwolter) [inline bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandler.java:158]: Please add a reasonable timeout like 10 sec.

File: bundles/org.openhab.binding.mystrom/src/main/resources/ESH-INF/thing/thing-types.xml [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+			<channel id="power" typeId="power-channel"/>
+			<channel id="temperature" typeId="temperature-channel"/>
+		</channels>
+
+		<config-description>
+			<parameter name="hostname" type="text">
+				<label>Hostname</label>
+				<description>The host name or IP address of the mystrom plug. It may or may not start with the
+					protocol, e.g. in
+					order to use https:// instead of the default http://.</description>
```
Comment (fwolter) [inline bundles/org.openhab.binding.mystrom/src/main/resources/ESH-INF/thing/thing-types.xml:22]: Is there a reason why you don't use https by default? You could remove the protocol here and add `<context>network-address</context>` to get a free validation, then.

File: bundles/org.openhab.binding.mystrom/src/main/resources/ESH-INF/thing/thing-types.xml [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+				<default>localhost</default>
+			</parameter>
+			<parameter name="refresh" type="integer">
```
Comment (fwolter) [inline bundles/org.openhab.binding.mystrom/src/main/resources/ESH-INF/thing/thing-types.xml:25]: You might want to specify a reasonable min value (e.g. 1). If the user specifys a 0 or a negative value, your code will crash.
```suggestion
			<parameter name="refresh" type="integer" unit="s">
```

File: bundles/org.openhab.binding.mystrom/src/main/resources/ESH-INF/thing/thing-types.xml [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xmlns:thing="https://openhab.org/schemas/thing-description/v1.0.0"
+	xsi:schemaLocation="https://openhab.org/schemas/thing-description/v1.0.0 https://openhab.org/schemas/thing-description-1.0.0.xsd">
+
+	<thing-type id="mystromplug">
+		<label>myStrom Smart Plug</label>
+		<description>Controls the myStrom smart plug</description>
+
+		<channels>
+			<channel id="switch" typeId="switch-channel"/>
```
Comment (fwolter) [inline bundles/org.openhab.binding.mystrom/src/main/resources/ESH-INF/thing/thing-types.xml:12]: You could use the system type. Then, you don't need to specify your own Channel type.
```suggestion
			<channel id="switch" typeId="system.power"/>
```

**Review by fwolter — CHANGES_REQUESTED** [review]
Your code looks good, only a few minor things.

There are some checkstyle warnings left, mainly concerning the readme. You could take a look at `target/code-analysis/report.html`.

There are some formatting issues. You can view them with `mvn spotless:check -Dspotless.check.skip=false` and fix them with `mvn spotless:apply`.

File: bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandler.java (hunk 7 above)
Comment (pail23) [reply bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandler.java]: changed to exception message

File: bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandler.java (hunk 8 above)
Comment (pail23) [reply bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandler.java]: done

File: bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandler.java (hunk 9 above)
Comment (pail23) [reply bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandler.java]: removed the log entry

File: bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandler.java (hunk 10 above)
Comment (pail23) [reply bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandler.java]: done

File: bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandler.java (hunk 11 above)
Comment (pail23) [reply bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandler.java]: removed the log entry

File: bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandler.java (hunk 12 above)
Comment (pail23) [reply bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandler.java]: added a 10 sec timeout

File: bundles/org.openhab.binding.mystrom/src/main/resources/ESH-INF/thing/thing-types.xml (hunk 13 above)
Comment (pail23) [reply bundles/org.openhab.binding.mystrom/src/main/resources/ESH-INF/thing/thing-types.xml]: done

File: bundles/org.openhab.binding.mystrom/src/main/resources/ESH-INF/thing/thing-types.xml (hunk 14 above)
Comment (pail23) [reply bundles/org.openhab.binding.mystrom/src/main/resources/ESH-INF/thing/thing-types.xml]: done

File: bundles/org.openhab.binding.mystrom/src/main/resources/ESH-INF/thing/thing-types.xml (hunk 15 above)
Comment (pail23) [reply bundles/org.openhab.binding.mystrom/src/main/resources/ESH-INF/thing/thing-types.xml]: done

**Comment by pail23** [comment]
@fwolter Thanks for the very helpful review comments. I have address them in the last comit. Please let me know in case I can further improve things.

File: bundles/pom.xml [hunk 16]
```diff
@@ -173,6 +173,7 @@
     <module>org.openhab.binding.mqtt.generic</module>
     <module>org.openhab.binding.mqtt.homeassistant</module>
     <module>org.openhab.binding.mqtt.homie</module>
+    <module>org.openhab.binding.mystrom</module>
```
Comment (Hilbrand) [inline bundles/pom.xml:176]:
```suggestion
    <module>org.openhab.binding.mystrom</module>
```

File: bundles/org.openhab.binding.mystrom/.project [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+		<filter>
+			<id>1599248758658</id>
+			<name></name>
+			<type>30</type>
+			<matcher>
+				<id>org.eclipse.core.resources.regexFilterMatcher</id>
+				<arguments>node_modules|.git|__CREATED_BY_JAVA_LANGUAGE_SERVER__</arguments>
+			</matcher>
+		</filter>
+	</filteredResources>
```
Comment (fwolter) [inline bundles/org.openhab.binding.mystrom/.project:27]: Can this be removed?

File: bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandler.java [hunk 18]
```diff
@@ -0,0 +1,166 @@
...
+
+    @Override
+    public void initialize() {
+        MyStromConfiguration config = getConfigAs(MyStromConfiguration.class);
+        if (config.hostname.startsWith(HTTP_REQUEST_URL_PREFIX)
+                || config.hostname.startsWith(HTTPS_REQUEST_URL_PREFIX)) {
+            this.hostname = config.hostname;
+        } else {
+            this.hostname = HTTP_REQUEST_URL_PREFIX + config.hostname;
+        }
```
Comment (fwolter) [inline bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandler.java:127]: Can this if be removed, then?

File: bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandler.java [hunk 19]
```diff
@@ -0,0 +1,166 @@
...
+
+        updateStatus(ThingStatus.UNKNOWN);
+        pollingJob = scheduler.scheduleWithFixedDelay(this::pollDevice, 0, config.refresh, TimeUnit.SECONDS);
+    }
+
+    @Override
+    public void dispose() {
+        super.dispose();
```
Comment (fwolter) [inline bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandler.java:135]: Base class destructors should be called at the end of the implementing destructors.

**Review by fwolter — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandler.java (hunk 18 above)
Comment (pail23) [reply bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandler.java:127]: Why would you remove this? And how would you handle the prefix? Always use http?

File: bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandler.java (hunk 19 above)
Comment (pail23) [reply bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandler.java]: done

File: bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandler.java (hunk 18 above)
Comment (fwolter) [reply bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandler.java:127]: As you defined `<context>network-address</context>` the user cannot enter an URL anymore, but a network address. So, the protocol should be suffixed always with your current implementation.

File: bundles/org.openhab.binding.mystrom/.project (hunk 17 above)
Comment (pail23) [reply bundles/org.openhab.binding.mystrom/.project]: done

File: bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandler.java (hunk 18 above)
Comment (pail23) [reply bundles/org.openhab.binding.mystrom/src/main/java/org/openhab/binding/mystrom/internal/MyStromHandler.java:127]: OK, I have adapted the code accordingly

**Comment by pail23** [comment]
Yes, because the plug itself only supports http. If you want to use https you would need to install a reverse proxy (which is quite complex for this case)

**Review by fwolter — APPROVED** [review]
LGTM

**Pull request merged**